### ISSUE TITLE
Sparrow to bitcoin-core connection problems

### ISSUE THREAD

**Issue by avatar4d** [sparrowwallet/sparrow#991]
Sparrow has been having trouble maintaining a connection to my private bitcoin-core node. It seems it can only connect to it after it has just been restarted. It also takes about 45s exactly every time when it does connect. Then after a random period of time (usually after only a few minutes) it will disconnect and never connect again unless bitcoind is restarted. 

I connected Specter to test if bitcoin-core was the issue, but Specter connects instantly and never loses a connection. It continues to work even while Sparrow is unable to reconnect. Attached is a Sparrow debug log from the point of startup until it flipped the connection toggle (note I had to split it up because GitHub  limits uploads to 25MB).

[xaa.log](https://github.com/sparrowwallet/sparrow/files/11771167/xaa.log)
[xab.log](https://github.com/sparrowwallet/sparrow/files/11771168/xab.log)
[xac.log](https://github.com/sparrowwallet/sparrow/files/11771169/xac.log)
[xad.log](https://github.com/sparrowwallet/sparrow/files/11771170/xad.log)
[xae.log](https://github.com/sparrowwallet/sparrow/files/11771172/xae.log)
[xaf.log](https://github.com/sparrowwallet/sparrow/files/11771174/xaf.log)
[xag.log](https://github.com/sparrowwallet/sparrow/files/11771175/xag.log)
[xah.log](https://github.com/sparrowwallet/sparrow/files/11771177/xah.log)
[xai.log](https://github.com/sparrowwallet/sparrow/files/11771178/xai.log)

**Comment by craigraw** [comment]
I have been looking into this, and think I understand the cause of this issue. In order to provide a fee rate histogram on the Send tab, Sparrow's internal Bitcoin Core server (called Cormorant) must handle the `mempool.get_fee_histogram` Electrum call. Unfortunately the Bitcoin Core RPC does not provide an efficient way to get this data, so all implementations (EPS/BWT/Cormorant etc) must use some form of the `getrawmempool` RPC call to get mempool fee information. I believe it is this call that is causing the issue on low powered machines now that we have a full mempool.

Can you test how long this call is taking on your node? e.g.
```
time bitcoin-cli getrawmempool true
```

I have already written an alternative approach to this call, which uses the call above on initialization only. Thereafter, it updates a local copy of the mempool fee data with non-verbose calls (`getrawmempool false`), followed by `getmempoolentry` on each new transaction added to the pool since the last call. 

However, I would like to determine if even the first call is unworkable on low powered nodes - your suggestion of "it seems it can only connect to it after it has just been restarted" implies this. Can you determine if your node is unable to handle `getrawmempool true` after it has been running for awhile?

**Comment by avatar4d** [comment]
0m46.37s real     0m12.20s user     0m10.89s system

**Comment by craigraw** [comment]
Great, that's confirmation it's the `getrawmempool true` call that's causing this.

Can you do the same with
```
time bitcoin-cli getrawmempool false
```

**Comment by avatar4d** [comment]
0m05.33s real     0m01.16s user     0m00.57s system

**Comment by craigraw** [comment]
Fixed in 171bf241. Cormorant now avoids calling `getrawmempool true` on slower systems, and initializes the fee rate histogram (loads mempool entries) in the background after a successful connection, thereafter updating its view of the mempool incrementally every minute. The fee rate chart on the Send tab is shown as soon as vsizes from all entries have been loaded.

**Comment by avatar4d** [comment]
Confirmed, thanks!
